Update jenkinsfile.adoc

c-okelly opened the pull request:
Add new section on how to set environment variables dynamically using scripts. Included link to section for more information on scripts.

bitwiseman commented:
@abayer  Does this look reasonable to you?

abayer commented:
I don't love it - that's definitely not something we want to be actively encouraging without a bunch of caveats. At a minimum, I'd suggest wrapping the `sh` calls in `"${...}"` - yeah, it's practically the same but I honestly don't think I intended for `FOO = sh ...` to pass validation, while `FOO = "${sh ...}"` definitely should pass validation (even if I'd rather it was never done, of course). And I think it's absolutely vital to point out that you can't do this with `agent none`.

c-okelly commented:
@abayer Thanks for your reply. I do agree it isn't very nice but I couldn't find a better alternative for dynamically assigning variables.

I think I must be misunderstanding your suggestion about using `${...}`.

When I try wrapping the sh section with `${...}` the pipeline throws the following error on startup.

`org.codehaus.groovy.control.MultipleCompilationErrorsException: startup failed: ...`

`EXIT_STATUS = ${sh(
                returnStatus: true,
                script: 'exit 1'
            )}`

For the section about pointing out for agent none I will add this into the documentation as follows.

`<1> An agent must be set at the top level of the pipeline. This will fail if agent if "agent none" is set `.

Let me know what you think and apologies if I have misunderstood the first part.

abayer commented:
@c-okelly - I mean `"${sh(...)}"` - note the double-quotes. =)

c-okelly commented:
@abayer Ahh sorry missed that. Should I use the """ """ in this case then since it is a multiline statement or convert it to a single line?

abayer commented:
@c-okelly Yeah.

c-okelly commented:
@bitwiseman Apologies for the number of commits. I noticed a few spelling mistakes and corrected them. I also added a note about trimming return values when using returnStdout.

bitwiseman commented:
@c-okelly More commits are no problem. Thanks for contributing.

abayer commented on content/doc/book/pipeline/jenkinsfile.adoc:
```diff
@@ -420,6 +420,45 @@ apply to all steps within the Pipeline.
+// Script //
+----
+<1> An agent must be set at the top level of the pipeline. This will fail if agent is set as `agent none`.
+<2> When using returnStdout a trailing whitespace will be append to the returned string. Use `.trim()` to remove this. 
```
Probably should be `returnStdout` (i.e., backticks)?

abayer commented on content/doc/book/pipeline/jenkinsfile.adoc:
```diff
@@ -420,6 +420,45 @@ apply to all steps within the Pipeline.
+}
+// Script //
+----
+<1> An agent must be set at the top level of the pipeline. This will fail if agent is set as `agent none`.
```
I'd backtick-quote agent in "if agent is set..."

abayer commented on content/doc/book/pipeline/jenkinsfile.adoc:
```diff
@@ -420,6 +420,45 @@ apply to all steps within the Pipeline.
 
+==== Setting environment variables dynamically 
+
+In the case where environment variable need to be set dynamically at run time this can be done with the use of a shell scripts (sh), Windows Batch Script (bat) or PowerShell Script (powershell). Each script can either returnStatus or returnStdout.  link:https://jenkins.io/doc/pipeline/steps/workflow-durable-task-step[More information on scripts].
```
I'd like it if the steps (`sh`, `bat`, `powershell`) and the parameter names (`returnStatus`, `returnStdout`) were backtick-quoted.

abayer commented on content/doc/book/pipeline/jenkinsfile.adoc:
```diff
@@ -420,6 +420,45 @@ apply to all steps within the Pipeline.
+
+In the case where environment variable need to be set dynamically at run time this can be done with the use of a shell scripts (sh), Windows Batch Script (bat) or PowerShell Script (powershell). Each script can either returnStatus or returnStdout.  link:https://jenkins.io/doc/pipeline/steps/workflow-durable-task-step[More information on scripts].
+
+Below is an example in a declarative pipeline using sh (shell) with both returnStatus and returnStdout.
```
Again, backtick-quoting, please? =)

c-okelly commented:
@bitwiseman  You beat me to it! Thanks
